benchmarks/distributed/dataloading.py

```diff
@@ -0,0 +1,363 @@
+        return x.gather(-2, index0).gather(-1, index1)
+
+
+class Collate(nn.Module):
+    def __init__(self, transform=None, device=None):
+        super().__init__()
+        self.transform = transform
+        self.device = device
+
+    def __call__(self, x: ImageNetData):
+        # move data to RAM
+        out = x.apply(lambda x: x.contiguous())
+        if self.device:
+            # move data to gpu
+            out = out.to(self.device)
+        if self.transform:
+            # apply transforms on gpu
+            out.images = self.transform(out.images)
+        return out
+
+
+@accept_remote_rref_udf_invocation
+class DummyTrainerNode:
+    def __init__(self, world_size) -> None:
+        print("DummyTrainerNode")
+        self.id = rpc.get_worker_info().id
+        self._prepare()
+        self.datanodes = []
+        self.world_size = world_size
+
+    def _prepare(self):
+        print("preparing data")
+        data_dir = Path("/datasets01_ontap/imagenet_full_size/061417/")
+        train_data_raw = datasets.ImageFolder(
+            root=data_dir / "train",
+            transform=transforms.Compose(
+                [transforms.Resize((256, 256)), transforms.PILToTensor()]
+            ),
+        )
+        train_data_raw.samples = train_data_raw.samples[
+            : len(train_data_raw) // FRACTION
+        ]
+
+        # val_transform = transforms.Compose(
+        #     [
+        #         transforms.Resize(256),
+        #         transforms.CenterCrop(224),
+        #         transforms.ToTensor(),
+        #         transforms.Normalize([0.485, 0.456, 0.406],
+        #                              [0.229, 0.224, 0.225]),
+        #     ]
+        # )
+
+        # val_data = datasets.ImageFolder(root=data_dir / "val", transform=val_transform)
+        # val_data.samples = val_data.samples[: len(val_data) // FRACTION]
+
+        train_data_tc = ImageNetData.from_dataset(train_data_raw)
+        # val_data_tc = ImageNetData.from_dataset(val_data)
+        self.data = train_data_tc
+
+    def train(self) -> None:
+        print("train")
+        len_data = self.data.shape[0]
+        pbar = tqdm.tqdm(total=len_data)
+
+        _prefetch_queue = collections.deque()
+        _last = 0
+        _next = BATCH_SIZE
+        for i in range(self.world_size - 1):
+            _prefetch_queue.append(
+                rpc.rpc_async(
+                    self.datanodes[i].owner(),
+                    DataNode.sample,
+                    args=(self.datanodes[i], range(_last, _next)),
+                )
+            )
+            _last = _next
+            _next = min(len_data - 1, _next + BATCH_SIZE)
+            pbar.update(BATCH_SIZE)
+
+        t0 = time.time()
+        iteration = 0
+        total = 0
+        while total < len_data:
+            batch = _prefetch_queue.popleft().wait()
+            i = iteration % (self.world_size - 1)
+            iteration += 1
+            _prefetch_queue.append(
```

Comment:
TODO: put this under a condition (we don't want to append if there is no data to collect)